include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1999 @@
+
+        struct skip_initialization_tag {};
+    }
+
+    namespace
+    {
+        tags::skip_initialization_tag  dont_init;
+    }
+
+    template <class T>
+    struct xtiny_concept
+    : public std::integral_constant<bool,
+                                    std::is_base_of<tags::xtiny_tag, std::decay_t<T>>::value>
+    {
+    };
+
+    /****************/
+    /* declarations */
+    /****************/
+
+    template <class VALUETYPE, index_t N, class REPRESENTATION>
+    class xtiny_impl;
+
+    template <class VALUETYPE, index_t N=runtime_size, class REPRESENTATION=void>
+    class xtiny;
+
+    /*********/
+    /* xtiny */
+    /*********/
+
+    /* Adds common functionality to the respective xtiny_impl */
+    template <class VALUETYPE, index_t N, class REPRESENTATION>
+    class xtiny
+    : public xtiny_impl<VALUETYPE, N, REPRESENTATION>
+    {
+      public:
+
+        using self_type = xtiny<VALUETYPE, N, REPRESENTATION>;
+        using base_type = xtiny_impl<VALUETYPE, N, REPRESENTATION>;
+        using value_type = typename base_type::value_type;
+        using const_value_type = typename base_type::const_value_type;
+        using reference = typename base_type::reference;
+        using const_reference = typename base_type::const_reference;
+        using pointer = typename base_type::pointer;
+        using const_pointer = typename base_type::const_pointer;
+        using iterator = typename base_type::iterator;
+        using const_iterator = typename base_type::const_iterator;
+        using reverse_iterator = typename base_type::reverse_iterator;
+        using const_reverse_iterator = typename base_type::const_reverse_iterator;
+        using size_type = typename base_type::size_type;
+        using difference_type = typename base_type::difference_type;
+        using index_type = typename base_type::index_type;
+
+        using base_type::owns_memory;
+        using base_type::has_fixed_size;
+        using base_type::static_size;
+
+        using base_type::base_type;
+
+        xtiny();
+        xtiny(xtiny const & rhs);
+        xtiny(xtiny && rhs);
+
+        template <class T, index_t M, class R>
+        explicit xtiny(xtiny<T, M, R> const & rhs);
+
+        template <class T, class A>
+        explicit xtiny(std::vector<T, A> const & v);
+
+        template <class T, std::size_t M>
+        explicit xtiny(std::array<T, M> const & v);
+
+        xtiny & operator=(xtiny const & rhs);
+        xtiny & operator=(xtiny && rhs);
+
+        xtiny & operator=(value_type const & v);
+
+        template <class T, class A>
+        xtiny & operator=(std::vector<T, A> const & v);
+
+        template <class T, std::size_t M>
+        xtiny & operator=(std::array<T, M> const & v);
+
+        template <class U, index_t M, class R>
+        xtiny & operator=(xtiny<U, M, R> const & rhs);
+
+        using base_type::assign;
+
+        template <class T>
+        void assign(std::initializer_list<T> v);
```

Comment:
The initializer list should only take value_type and not template T, otherwise implicit conversion costs 2ns :) I spent a day wondering about a benchmark and it was that, hehe.

Reply:
Having it templated on T is important to avoid type conversion warnings in code like this:
```c++
xtiny<unsigned, 3> vi{1, 2, 3};    // T = int
xtiny<float, 3> vf{1.0, 2.0, 3.0}; // T = double
```
However, we can provide an overload with `initializer_list<value_type>`. Or does the compiler automatically convert the initializer list arguments?

Reply:
I'm pretty sure that std::vector etc. do not have a templated initializer_list constructor, so I don't think we should have one. And the conversion warnings in this case would be legit, so I think it's fine to just remove it.

In my test, the compiler basically used the initializer_list<int>() overload, and then did the type conversion upon assignment (instead of statically at compile time) and that costs time in the benchmark.

Reply:
Dropping the templated variant would be very unfortunate:
```c++
template <class REAL>
void foo(xtiny<REAL, 2> & v)
{
    v.assign({-0.5, 0.5});  // doesn't compile when REAL=float

    v.assign({static_cast<REAL>(-0.5), static_cast<REAL>(0.5)}); // UGLY!!
}
```

Reply:
that's really weird. E.g. this compiles fine with clang:

```
std::vector<float> a({1.,2.,3.});
```

Is it a compiler setting or MSVC?